### PR TITLE
refactor TableScanExecutor and IndexScanExecutor

### PR DESCRIPTION
**Pull request by niedhui**
Signed-off-by: niedhui <niedhui@gmail.com>

## What have you changed? (mandatory)

1. refactor `TableScanExecutor` and `IndexScanExetuor`,  since the logic when they act as a `ScanExecutor` is nearly the same, except when got a row , they use different strategy to populate the  `Row::Origin`. so I extract a `InnerExecutor` to do the different work.
if we want add some logic on table scan and index scan later( like push down the deadline to the bottom executor or add more metrics), we only need do it  on  `ScanExecutor`

2. *Question*: for index scan , in `next`, is use `self.unique && util::is_point(range)` to check  whether the range is a point, but on `start_scan` and `stop_scan`, it only use `util::is_point(range)`,  is it correct?

and *Feel Free* to close this PR if it is unnecessary(these codes is written when I was trying to figure out how  table scan and index scan is work)

## What are the type of the changes? (mandatory)

Improvement

## How has this PR been tested? (mandatory)

unit test

## Does this PR affect documentation (docs) update? (mandatory)

No

## Does this PR affect tidb-ansible update? (mandatory)

No

### CODE REVIEW

**Comment by siddontang** [comment]
Amazing, thanks @niedhui

**Comment by breeswish** [comment]
Thanks for your great work. Looks like this PR makes the code more complicated in order to reduce duplications. In fact the common part for TableScan / IndexScan is that they both follow a same way scanning rows from point ranges or non-point ranges. Then they will act differently when receive the row. How about just provide different logic for the process of receiving the row?

For Question 2, you are right. It is incorrect. Feel free to fix it.

**Comment by niedhui** [comment]
@breeswish 
Although these codes show me the differences between a table-scan and an index-scan clearly , but do introduce another layer( complicated).

first, let's put the test codes aside, they are just some helper functions move from one place to anoter.

and:
> In fact the common part for TableScan / IndexScan is that they both follow a same way scanning rows from point ranges or non-point ranges. 

this is what the `ScanExecutor` does,  `next` is same, `start_scan` is same, and so other methods.

>  they will act differently when receive the row. How about just provide different logic for the process of receiving the row?

this is what the `decode_row` of `InnerExecutor` does.

**Comment by breeswish** [comment]
@niedhui Yeah, I know your point. I was just feeling that the generic over Table/Index (InnerExecutor) is not very common (common logic outside, inner executor place specialized logics). **Maybe** better to put common logics inside and put specialized logics outside.

**Comment by niedhui** [comment]
@breeswish 
>  Maybe better to put common logics inside and put specialized logics outside.

this is what I tried first, but the `common logic` depneds on some many variables(almost every field of TableScanExecutor/IndexScanExecutor) , and there was a `Scanner` handling some common logic already...,  so I gave up 🤣 

Let's close this PR and leave it as it is. I'll fix the `is_point` issue of IndexScanExecutor in a separate PR, in case it is forgotten.

**Comment by AndreMouche** [comment]
For question 2, since there are two types of index data: unique index && non-unique index, and for the non-unique index the `util::is_point` may return true while is not a point request. So we do need `self.unique && util::is_point(range)` to check whether the range is a point. @breeswish @niedhui @hanfei1991

**Comment by breeswish** [comment]
I tried with some other ways to extract common code, looks like your idea is the best solution. I prefer to further improve your PR. I suggest to create `struct TableScanExecutor(ScanExecutor<..>)` instead of make a type alias, so that there will be less confusing.

**Comment by niedhui** [comment]
if using the tuple struct to create `struct TableScanExecutor(ScanExecutor<..>)` ,  we need to impl `Executor` trait on `TableScanExecutor`, intead of `ScanExecutor`? since what we need is an `Executor`

**Comment by breeswish** [comment]
Okay, there are a lot of functions to implement to meet the Executor interface. Let's stick to this implementation.

**Comment by breeswish** [comment]
@AndreMouche PTAL

File: src/coprocessor/dag/executor/table_scan.rs [hunk 1]
```diff
@@ -330,7 +191,7 @@ mod tests {
         let (snapshot, start_ts) = wrapper.store.get_snapshot();
         let store = SnapshotStore::new(snapshot, start_ts, IsolationLevel::SI, true);
         let mut table_scanner =
-            TableScanExecutor::new(wrapper.table_scan, wrapper.ranges, store, false).unwrap();
+            TableScanExecutor::table_scan(wrapper.table_scan, wrapper.ranges, store, true).unwrap();
```
Comment (AndreMouche) [inline src/coprocessor/dag/executor/table_scan.rs]: I prefer to use `TableScanExecutor::new` to create a `TableScanExecutor`

Comment (niedhui) [reply src/coprocessor/dag/executor/table_scan.rs]: `TableScanExecutor ` and `IndexScanExecutor` is a type alias for `ScanExecutor`,  and the `new` method is used for create either of these, so I add `table_scan` and `index_scan`, which using `new` internally

File: src/coprocessor/dag/executor/index_scan.rs [hunk 2]
```diff
@@ -148,131 +71,54 @@ impl<S: Store> IndexScanExecutor<S> {
...
-    }
-
-    fn new_scanner(&self, range: KeyRange) -> Result<Scanner<S>> {
-        // Since the unique index wouldn't always come with
-        // self.unique = true. so the key-only would always be false.
-        Scanner::new(&self.store, ScanOn::Index, self.desc, false, range).map_err(Error::from)
+        Ok(Some(Row::origin(handle, values, columns)))
     }
 
     fn is_point(&self, range: &KeyRange) -> bool {
```
Comment (breeswish) [inline src/coprocessor/dag/executor/index_scan.rs]: Add #[inline]

File: src/coprocessor/dag/executor/index_scan.rs [hunk 3]
```diff
@@ -148,131 +71,54 @@ impl<S: Store> IndexScanExecutor<S> {
...
-        }
 
-        if let Some(range) = self.key_ranges.peek() {
-            if !self.desc {
-                self.scan_range.set_start(range.get_start().to_owned());
-            } else {
-                self.scan_range.set_end(range.get_end().to_owned());
-            }
-        }
+    fn scan_on(&self) -> ScanOn {
```
Comment (breeswish) [inline src/coprocessor/dag/executor/index_scan.rs]: Add #[inline]

File: src/coprocessor/dag/executor/index_scan.rs [hunk 4]
```diff
@@ -148,131 +71,54 @@ impl<S: Store> IndexScanExecutor<S> {
...
-    }
-
-    fn collect_output_counts(&mut self, counts: &mut Vec<i64>) {
-        if let Some(cur_counts) = self.counts.as_mut() {
-            counts.append(cur_counts);
-            cur_counts.push(0);
-        }
+    // Since the unique index wouldn't always come with
+    // self.unique = true. so the key-only would always be false.
+    fn key_only(&self) -> bool {
```
Comment (breeswish) [inline src/coprocessor/dag/executor/index_scan.rs]: Add #[inline]

File: src/coprocessor/dag/executor/scan.rs [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+use std::{iter::Peekable, mem, sync::Arc, vec::IntoIter};
+
+use super::{Executor, ExecutorMetrics, Row, ScanOn, Scanner};
+use crate::coprocessor::codec::table;
+use crate::coprocessor::{Error, Result};
+use crate::storage::{Key, Store};
+use kvproto::coprocessor::KeyRange;
+use tipb::schema::ColumnInfo;
+
+pub trait InnerExecutor {
```
Comment (breeswish) [inline src/coprocessor/dag/executor/scan.rs]: Please add comments for public traits and its interfaces

File: src/coprocessor/dag/executor/scan.rs [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+        key: Vec<u8>,
+        value: Vec<u8>,
+        columns: Arc<Vec<ColumnInfo>>,
+    ) -> Result<Option<Row>>;
+    fn is_point(&self, range: &KeyRange) -> bool;
+    fn scan_on(&self) -> ScanOn;
+    fn key_only(&self) -> bool;
+}
+
+// executor for table scan and index scan
```
Comment (breeswish) [inline src/coprocessor/dag/executor/scan.rs]: s/executor/Executor

**Comment by AndreMouche** [comment]
Please merge the master and fix the conflicts @niedhui

**Comment by niedhui** [comment]
@AndreMouche PTAL

**Comment by ngaut** [comment]
Great work.

**Pull request merged**